Cython/Compiler/Optimize.py

```diff
@@ -1991,10 +1991,18 @@ def _handle_simple_function_frozenset(self, node, pos_args):
         """
         if len(pos_args) != 1:
             return node
+        # If args is an empty sequence, remove it
         if pos_args[0].is_sequence_constructor and not pos_args[0].args:
             del pos_args[0]
-        elif isinstance(pos_args[0], ExprNodes.ListNode):
+            return node
+        if isinstance(pos_args[0], ExprNodes.ListNode):
             pos_args[0] = pos_args[0].as_tuple()
+        # Currently, if a tuple is constant, a permanent tuple would be created
+        # Therefore, we convert this tuple to list, only keeping a permanent frozenset
+        # See https://github.com/cython/cython/pull/4098#issuecomment-834234964
+        if isinstance(pos_args[0], ExprNodes.TupleNode) and \
+            not isinstance(pos_args[0].constant_result, ExprNodes.NotConstant):
+                pos_args[0] = pos_args[0].as_list()
```

Comment:
This should go before the `as_tuple()` call above in order to avoid superfluous back and forth conversion.
```suggestion
        # Avoid creating a tuple constant when we only need it once to create the frozenset.
        if isinstance(pos_args[0], ExprNodes.TupleNode) and pos_args[0].has_constant_result():
                pos_args[0] = pos_args[0].as_list()
```